### ISSUE TITLE
MERGE should raise when multiple source rows match target row

### ISSUE THREAD

**Issue by mrjsj** [delta-io/delta-rs#2407]
# Environment

**Delta-rs version**: 0.16.4

**Binding**: Python 0.16.4

**Environment**:
- **Cloud provider**: None
- **OS**: MacOS Sonoma 14.4
- **Other**:

***
# Bug

**What happened**:
TableMerger.when_matched_update_all() inserts records when there is a match on multiple source columns
TableMerger.when_matched_update() inserts records when there is a match on multiple source columns

**What you expected to happen**:
TableMerger.when_matched_update_all() should throw an error if a target record matches multiple source records 
TableMerger.when_matched_update() should throw an error if a target record matches multiple source records 

They should never insert new records.

**How to reproduce it**:
Using polars 0.20.7
```python
import polars as pl
from deltalake import DeltaTable

base_df = pl.DataFrame(
    {
        "id": [1, 2],
        "attr": ["x", "y"]
    }
)

base_df.write_delta("./test_cdc", mode="overwrite")

dt = DeltaTable("./test_cdc")

print(pl.DataFrame(dt.to_pyarrow_table()))

cdc_df = pl.DataFrame(
    {
        "id": [1,1,1,2,2],
        "attr": ["a","b","c","d","e"],
        "op": ["U", "U", "U", "U", "U"]
    }
)


(
    dt.merge(
        cdc_df.to_arrow(),
        "s.id = t.id",
        source_alias="s",
        target_alias="t",
    )
    .when_matched_update(
        updates={"t.attr": "s.attr"},
        predicate="s.op = 'U'")
    .execute()
)

print(pl.DataFrame(dt.to_pyarrow_table()))
```

Gives the following output

**base table**
```bash
shape: (2, 2)
┌─────┬──────┐
│ id  ┆ attr │
│ --- ┆ ---  │
│ i64 ┆ str  │
╞═════╪══════╡
│ 1   ┆ x    │
│ 2   ┆ y    │
└─────┴──────┘
```

**After TableMerger is executed**
```bash
shape: (5, 2)
┌─────┬──────┐
│ id  ┆ attr │
│ --- ┆ ---  │
│ i64 ┆ str  │
╞═════╪══════╡
│ 2   ┆ d    │
│ 2   ┆ e    │
│ 1   ┆ a    │
│ 1   ┆ b    │
│ 1   ┆ c    │
└─────┴──────┘
```
**More details**:
Info on the specific case in slack: https://delta-users.slack.com/archives/C013LCAEB98/p1712673309723829

**Comment by ion-elgreco** [comment]
@mrjsj were you still working on this? I remember you had some fixes ready

**Comment by mrjsj** [comment]
Hi @ion-elgreco 
I did try, but got stuck for some reason. My "progress" is here: https://github.com/delta-io/delta-rs/compare/main...mrjsj:delta-rs:fix/raise-if-source-has-duplicates

I'm not really confident in Rust

**Comment by mrjsj** [comment]
@ion-elgreco  i have worked a little more on this; in a new branch. Rust tests run fine, however Python unit tests are failing when `streaming_exec=True`.

the `execute()` never merges to the table. It happens when I `.collect().await?` on a DataFrame used for duplication detection (see link).

Any ideas how to solve this?

https://github.com/mrjsj/delta-rs/blob/23eebf9adb9c21967c93c92796f78d7c1d2fe906/crates/core/src/operations/merge/mod.rs#L928C1-L944C17

**Comment by ion-elgreco** [comment]
> @ion-elgreco  i have worked a little more on this; in a new branch. Rust tests run fine, however Python unit tests are failing when `streaming_exec=True`.
> 
> the `execute()` never merges to the table. It happens when I `.collect().await?` on a DataFrame used for duplication detection (see link).
> 
> Any ideas how to solve this?
> 
> https://github.com/mrjsj/delta-rs/blob/23eebf9adb9c21967c93c92796f78d7c1d2fe906/crates/core/src/operations/merge/mod.rs#L928C1-L944C17

That's because in streaming a RecordBatchStream can only be collected once, so first you collect it do your matchings checks, but when you write the table, there are no more incoming recordbatches.

You might need to look into extension nodes that can keep a state

**Comment by jun-su-lee** [comment]
Hi,

I saw the previous investigation and partial fixes by @mrjsj — really appreciate the effort there.

If this issue is still open, I’d be happy to take a look and see whether I can help, possibly in collaboration or by exploring an alternative approach.

Please let me know if that would be appropriate. I want to make sure I’m not duplicating or interfering with existing work.

Thanks!

**Comment by mrjsj** [comment]
Hi @kiwwwwwi 
I'm not currently working on this, so feel free to take a jab.

**Comment by jun-su-lee** [comment]
take

**Comment by adampolomski** [comment]
@kiwwwwwi let me know if you’d like any help with this. I’m more than happy to take a look if useful.

**Comment by adampolomski** [comment]
@ion-elgreco would it be OK if I took this one?

**Comment by ion-elgreco** [comment]
@adampolomski Yes go ahead! If you want to do an issue, you can say `take`in the comment

**Comment by adampolomski** [comment]
take
